application/apps/indexer/session/src/state/session_file.rs

```diff
@@ -0,0 +1,279 @@
+        &mut self,
+        source_id: u8,
+        state_cancellation_token: CancellationToken,
+    ) -> Result<SessionFileState, NativeError> {
+        if self.writer.is_none() {
+            return Err(NativeError {
+                severity: Severity::ERROR,
+                kind: NativeErrorKind::Grabber,
+                message: Some(String::from("Session file isn't assigned yet")),
+            });
+        }
+        if !self.has_updates {
+            return Ok(SessionFileState::NoChanges);
+        }
+        let writer = &mut (self.writer.as_mut().ok_or(NativeError {
+            severity: Severity::ERROR,
+            kind: NativeErrorKind::Grabber,
+            message: Some(String::from("Session file writer wasn't created")),
+        })?);
+        writer.flush().map_err(|e| NativeError {
+            severity: Severity::ERROR,
+            kind: NativeErrorKind::Io,
+            message: Some(e.to_string()),
+        })?;
+        self.last_message_timestamp = Instant::now();
+        self.has_updates = false;
+        self.update(source_id, state_cancellation_token).await
+    }
+
+    pub async fn update(
+        &mut self,
+        source_id: u8,
+        state_cancellation_token: CancellationToken,
+    ) -> Result<SessionFileState, NativeError> {
+        let grabber = &mut (self.grabber.as_mut().ok_or(NativeError {
+            severity: Severity::ERROR,
+            kind: NativeErrorKind::Grabber,
+            message: Some(String::from("Grabber isn't inited")),
+        })?);
+        let prev = grabber.log_entry_count().unwrap_or(0) as u64;
+        if let Some(range) = grabber.update_from_file(Some(state_cancellation_token))? {
+            self.observing.add_range(range, source_id);
+        }
+        let current = grabber.log_entry_count().unwrap_or(0) as u64;
+        Ok(if prev != current {
+            SessionFileState::Changed
+        } else {
+            SessionFileState::NoChanges
+        })
+    }
+
+    pub fn grab(&self, range: &LineRange) -> Result<Vec<GrabbedElement>, NativeError> {
+        let grabber = &mut (self.grabber.as_ref().ok_or(NativeError {
+            severity: Severity::ERROR,
+            kind: NativeErrorKind::Grabber,
+            message: Some(String::from("Grabber isn't inited")),
+        })?);
+        let rows = grabber.grab_content(range).map_err(|e| NativeError {
+            severity: Severity::ERROR,
+            kind: NativeErrorKind::Grabber,
+            message: Some(format!("{e}")),
+        })?;
+        let mapped_ranges = self.observing.get_mapped_ranges(&range.range);
+        let from = *range.range.start() as usize;
+        Ok(rows
+            .into_iter()
+            .enumerate()
+            .map(|(i, content)| {
+                let pos = from + i;
+                GrabbedElement {
+                    source_id: mapped_ranges.source(pos as u64).unwrap_or(0),
+                    content,
+                    pos,
+                    row: pos,
```

Comment:
but is that really correct? `pos` was not the same as `row` AFAIR

Reply:
yes, here it's correct:
- main view (listing of session file) `pos == position_in_session_file`  && `pos == row`
- search results view (listing search results) `pos == position_in_session_file` && `row == position_in_search_results_view`